## Notebook 7 – Loading Models Manually


## Loading Models Manually in Hugging Face

Loading a model manually means instead of using the simple pipeline(), we directly load the tokenizer and model using AutoTokenizer and AutoModel....

Simple difference

| Approach       | Example                       | Use                         |
| -------------- | ----------------------------- | --------------------------- |
| `pipeline()`   | High-level API                | Easy, quick inference       |
| Manual loading | `AutoTokenizer` + `AutoModel` | More control over the model |


In [1]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

# Model name from Hugging Face
model_name = "distilbert-base-uncased-finetuned-sst-2-english"

# 1. Load tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name)

# 2. Load pre-trained model
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# 3. Input text
text = "I love this product!"

# 4. Tokenize
inputs = tokenizer(
    text,
    return_tensors="pt"
)

# 5. Send input to model
outputs = model(**inputs)

# 6. Get logits
logits = outputs.logits

print(logits)

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

C:\Users\new\anaconda3\Lib\site-packages\huggingface_hub\file_download.py:149: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\new\.cache\huggingface\hub\models--distilbert-base-uncased-finetuned-sst-2-english. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

C:\Users\new\anaconda3\Lib\site-packages\huggingface_hub\file_download.py:149: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\new\.cache\huggingface\hub. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors: reconstructing file:   0%|          |  0.00B /  268MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

tensor([[-4.3599,  4.7156]], grad_fn=<AddmmBackward0>)


In Hugging Face, loading a model manually means directly loading the tokenizer and pre-trained model instead of using the high-level `pipeline()` API.

We commonly use `AutoTokenizer.from_pretrained()` to load the tokenizer and `AutoModel` or a task-specific class such as `AutoModelForSequenceClassification.from_pretrained()` to load the pre-trained model and its learned weights.

The tokenizer converts text into inputs such as `input_ids` and `attention_mask`, and the model processes these inputs to produce outputs such as logits or predictions.

The basic flow is:

**Text → Tokenizer → Token IDs + Attention Mask → Model → Output**

Manual loading gives us more control over tokenization, model inputs, outputs, and inference compared with a pipeline.


| Pipeline                               | Manual                               |
| -------------------------------------- | ------------------------------------ |
| `pipeline()` does things automatically | We do things step-by-step            |
| Easy                                   | More control                         |
| Less code                              | More code                            |
| Good for beginners                     | Good for understanding/customization |


Manual model loading means loading the tokenizer and the pre-trained model separately using Hugging Face.

The tokenizer converts text into token IDs, and the model processes those token IDs to produce the output.

We commonly use `AutoTokenizer.from_pretrained()` for the tokenizer and `AutoModel.from_pretrained()` or a task-specific model class for the model.

**Simple flow:**

Text → Tokenizer → Token IDs → Model → Output


In [2]:
## Auto Loading using pipeline()

# Here, Hugging Face handles the tokenizer and model for us.

from transformers import pipeline

# Create a sentiment analysis pipeline
# Hugging Face automatically loads:
# 1. Tokenizer
# 2. Pre-trained model
# 3. Prediction process

classifier = pipeline("sentiment-analysis")

# Give text to the pipeline
text = "I love this movie!"

# Get prediction
result = classifier(text)

print(result)

[transformers] No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f.
Using a pipeline without specifying a model name and revision in production is not recommended.


Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

[{'label': 'POSITIVE', 'score': 0.9998775720596313}]


In [3]:
## Manual load 

from transformers import AutoTokenizer
from transformers import AutoModelForSequenceClassification
import torch

# Name of the pre-trained model
model_name = "distilbert-base-uncased-finetuned-sst-2-english"

# ------------------------------------------------
# STEP 1: Load the tokenizer
# ------------------------------------------------
# Tokenizer converts text into token IDs
tokenizer = AutoTokenizer.from_pretrained(model_name)


# ------------------------------------------------
# STEP 2: Load the pre-trained model
# ------------------------------------------------
# Model contains the already-trained weights
model = AutoModelForSequenceClassification.from_pretrained(model_name)


# ------------------------------------------------
# STEP 3: Give input text
# ------------------------------------------------
text = "I love this movie!"


# ------------------------------------------------
# STEP 4: Convert text into model inputs
# ------------------------------------------------
# return_tensors="pt" means return PyTorch tensors
inputs = tokenizer(
    text,
    return_tensors="pt"
)

print("Model Inputs:")
print(inputs)


# ------------------------------------------------
# STEP 5: Give inputs to the model
# ------------------------------------------------
# The model produces raw scores called logits
outputs = model(**inputs)

print("Logits:")
print(outputs.logits)


# ------------------------------------------------
# STEP 6: Convert logits into probabilities
# ------------------------------------------------
probabilities = torch.softmax(outputs.logits, dim=-1)

print("Probabilities:")
print(probabilities)


# ------------------------------------------------
# STEP 7: Find the class with highest probability
# ------------------------------------------------
prediction = torch.argmax(probabilities, dim=-1)

print("Predicted class ID:")
print(prediction.item())


# ------------------------------------------------
# STEP 8: Convert class ID into label
# ------------------------------------------------
labels = ["NEGATIVE", "POSITIVE"]

print("Prediction:")
print(labels[prediction.item()])

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

Model Inputs:
{'input_ids': tensor([[ 101, 1045, 2293, 2023, 3185,  999,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1]])}
Logits:
tensor([[-4.3246,  4.6837]], grad_fn=<AddmmBackward0>)
Probabilities:
tensor([[1.2238e-04, 9.9988e-01]], grad_fn=<SoftmaxBackward0>)
Predicted class ID:
1
Prediction:
POSITIVE


| Stage / Term                 | Simple Meaning                                    | Input                         | Output                        |
| ---------------------------- | ------------------------------------------------- | ----------------------------- | ----------------------------- |
| **Text**                     | Original human-readable sentence                  | `"I love this movie"`         | Text                          |
| **Tokenizer loading**        | Loads the tokenizer associated with the model     | Model name                    | Tokenizer                     |
| **AutoTokenizer**            | Automatically selects the correct tokenizer       | Model name                    | Tokenizer object              |
| **Tokenization**             | Breaks text into tokens                           | Text                          | Tokens                        |
| **Token IDs**                | Converts tokens into numbers                      | Tokens                        | `[101, 1045, ...]`            |
| **Configuration**            | Contains model settings                           | Model/config file             | Model configuration           |
| **Model loading**            | Loads the pre-trained model and weights           | Model name                    | Model                         |
| **AutoModel**                | Automatically selects the base model architecture | Model name                    | Model object                  |
| **Task-specific Auto class** | Loads a model designed for a particular task      | Model name                    | Task-specific model           |
| **Model inputs**             | Information given to the model                    | `input_ids`, `attention_mask` | Tensors                       |
| **Model output**             | Raw result produced by model                      | Model inputs                  | `logits`, hidden states, etc. |
| **Post-processing**          | Converts raw output into understandable result    | Logits                        | Label/probability             |
| **Final prediction**         | Final answer                                      | Processed output              | `POSITIVE`, `NEGATIVE`, etc.  |


| Concept                        | Simple Meaning                                                       | Example / Code                                          | Output / Purpose                     |
| ------------------------------ | -------------------------------------------------------------------- | ------------------------------------------------------- | ------------------------------------ |
| **Tokenizer Loading**          | Loads the tokenizer needed to convert text into tokens and token IDs | `tokenizer = AutoTokenizer.from_pretrained(model_name)` | Tokenizer object                     |
| **Model Loading**              | Loads the pre-trained model and its learned weights                  | `model = AutoModel.from_pretrained(model_name)`         | Model object                         |
| **Configuration**              | Contains the model's settings/architecture information               | `config = AutoConfig.from_pretrained(model_name)`       | Model configuration                  |
| **Model Inputs**               | Information given to the model after tokenization                    | `input_ids`, `attention_mask`                           | Tensors processed by model           |
| **Model Outputs**              | Results produced by the model                                        | `outputs = model(**inputs)`                             | Logits, hidden states, etc.          |
| **AutoTokenizer**              | Automatically selects the correct tokenizer for a model              | `AutoTokenizer.from_pretrained()`                       | Handles text → tokens/IDs            |
| **AutoModel**                  | Automatically selects and loads the base Transformer model           | `AutoModel.from_pretrained()`                           | Base model representation            |
| **Task-specific Auto Classes** | Automatically loads a model designed for a particular task           | `AutoModelForSequenceClassification`                    | Classification, generation, QA, etc. |


In [ ]:
Tokenizer loading
Model loading
Configuration
Model inputs
Model outputs
AutoTokenizer
AutoModel
Task-specific Auto classes

Understand:
Text
 ↓
Tokenizer
 ↓
Token IDs
 ↓
Model
 ↓
Model Output
 ↓
Post-processing
The implementation should show the intermediate values wherever practical.
